From c34888d472f500045b55724efcf8d9fa380a030d Mon Sep 17 00:00:00 2001
From: John Wang <wangzhiqiang02@inspur.com>
Date: Mon, 16 Jan 2023 11:04:51 +0800
Subject: [PATCH 1/1] Do not truncate log files after reboot

The log file will be truncated during the log_init process, so that the
logs will be lost after bmc(obmc-console) restarts.

Here we change the logic to loop through the sol logs instead of
truncating them when they already exist in the bmc system

Signed-off-by: John Wang <wangzhiqiang02@inspur.com>
Change-Id: I9a14867c7ec8bfb6f25edcc8f383afd44f8ee3ad
---
 log-handler.c | 40 ++++++++++++++++++++++++++++++++--------
 1 file changed, 32 insertions(+), 8 deletions(-)

diff --git a/log-handler.c b/log-handler.c
index a545505..3a0632b 100644
--- a/log-handler.c
+++ b/log-handler.c
@@ -24,11 +24,13 @@
 #include <unistd.h>
 
 #include <sys/mman.h>
-
+#include <sys/types.h>
 #include <linux/types.h>
 
 #include "console-server.h"
 
+#define LOGFILE_O_FLAGS (O_WRONLY | O_CREAT | O_APPEND)
+
 struct log_handler {
 	struct handler handler;
 	struct console *console;
@@ -62,7 +64,7 @@ static int log_trim(struct log_handler *lh)
 		/* don't return, as we need to re-open the logfile */
 	}
 
-	lh->fd = open(lh->log_filename, O_RDWR | O_CREAT | O_TRUNC, 0644);
+	lh->fd = open(lh->log_filename, LOGFILE_O_FLAGS, 0644);
 	if (lh->fd < 0) {
 		warn("Can't open log buffer file %s", lh->log_filename);
 		return -1;
@@ -126,6 +128,30 @@ static enum ringbuffer_poll_ret log_ringbuffer_poll(void *arg, size_t force_len
 	return RINGBUFFER_POLL_OK;
 }
 
+static int log_create(struct log_handler *lh)
+{
+	off_t pos;
+
+	lh->fd = open(lh->log_filename, LOGFILE_O_FLAGS, 0644);
+	if (lh->fd < 0) {
+		warn("Can't open log buffer file %s", lh->log_filename);
+		return -1;
+	}
+	pos = lseek(lh->fd, 0, SEEK_CUR);
+	if (pos < 0) {
+		warn("Can't query log position for file %s", lh->log_filename);
+		close(lh->fd);
+		return -1;
+	}
+
+	if ((size_t)pos >= lh->maxsize) {
+		return log_trim(lh);
+	}
+
+	lh->size = pos;
+	return 0;
+}
+
 static int log_init(struct handler *handler, struct console *console,
 		    struct config *config)
 {
@@ -155,12 +181,6 @@ static int log_init(struct handler *handler, struct console *console,
 		filename = default_filename;
 	}
 
-	lh->fd = open(filename, O_RDWR | O_CREAT | O_TRUNC, 0644);
-	if (lh->fd < 0) {
-		warn("Can't open log buffer file %s", filename);
-		return -1;
-	}
-
 	lh->log_filename = strdup(filename);
 
 	rc = asprintf(&lh->rotate_filename, "%s.1", filename);
@@ -169,6 +189,10 @@ static int log_init(struct handler *handler, struct console *console,
 		return -1;
 	}
 
+	if (log_create(lh) < 0) {
+		return -1;
+	}
+
 	lh->rbc = console_ringbuffer_consumer_register(console,
 						       log_ringbuffer_poll, lh);
 
-- 
2.43.2

